ocaml/typing/solver.ml

```diff
@@ -807,7 +836,107 @@ module Solver_mono (C : Lattices_mono) = struct
     let lower = Option.value lower ~default:(C.min obj) in
     let vlower = Option.value vlower ~default:[] in
     let vupper = Option.value vupper ~default:[] in
-    { level; upper; lower; vlower; vupper; id }
+    let optcopy = None in
+    { level; upper; lower; vlower; vupper; id; optcopy }
+
+
+  type varcond = { b : 'a. 'a var -> bool; } [@@unboxed]
+
+  let rec copy_under_condition_v
+      : type a. copylog:_ -> varcond -> ?current_level:int ->
+        a C.obj -> a var -> a var =
+    fun ~copylog cond ?current_level obj v ->
+      if not (cond.b v) then v
+      else begin
+        match v.optcopy with
+        | Some v' -> v'
+        | None ->
+          let copy = fresh ?upper:(Some v.upper) ?lower:(Some v.lower) obj in
+          set_optcopy ~changes:copylog v (Some copy);
+          let vupper = List.map
+            (fun (Amorphvar (u, f)) ->
+              let src = C.src obj f in
+              let ucopy = copy_under_condition_v ~copylog cond ?current_level src u in
+              (Amorphvar (ucopy, f)))
+            v.vupper;
+          in
+          let vlower = List.map
+            (fun (Amorphvar (u, f)) ->
+              let src = C.src obj f in
+              let ucopy = copy_under_condition_v ~copylog cond ?current_level src u in
+              (Amorphvar (ucopy, f)))
+            v.vlower
+          in
+          copy.level <- v.level;
+          copy.vupper <- vupper;
+          copy.vlower <- vlower;
+          Option.iter (fun level -> update_level_v ~log:None obj level copy) current_level;
```

Comment:
The semantics of `current_level` as a parameter appear to be that it places an optional upper bound on the level. This makes sense in the `copy_generic` case, but it's not obvious it's the right semantics for the `copy_all` case. Which values of `current_level` do we actually end up using for `copy_all` once we start using it in the type-checker?

Reply:
Indeed, copy_all should take `None` for the optional `current_level` parameter. I changed the exposed signatures in `mode_intf.mli` to enforce this.